Scale q

kgori opened the pull request:
# A clear and descriptive title (No issue numbers please)

Scale Q matrix so that any branch lengths calculated using the Q will be in units of expected substitutions per site.

* [x] :sparkles: New feature (A non-breaking change which adds functionality).
* [ ] :bug: Bug fix (A non-breaking change, which fixes an issue).
* [ ] :boom: Breaking change (fix or feature that would cause existing functionality to change).

## :clipboard: Additional detail

- If you have implemented new features or behaviour
  - **Provide a description of the addition** in as many details as possible.
  Entries in the Q matrix are scaled so that model frequencies * -trace(Q) = 1.

  - **Provide justification of the addition**.
  This scaling is standard practice in phylogenetic inference ([wikipedia link](https://en.wikipedia.org/wiki/Models_of_DNA_evolution#Scaling_of_branch_lengths)).

- If you have changed current behaviour...
  - **Describe the behaviour prior to you changes**
  P and Q matrices will lead to inconsistent interpretations of branch length, not comparable between models.

  - **Describe the behaviour after your changes** and justify why you have made the changes,
    Please describe any breakages you anticipate as a result of these changes.
  Branch lengths derived from the models are all on the same scale - a branch length of 1 means 1 expected substitution per site.

  - **Does your change alter APIs or existing exposed methods/types?**
    No

  - If you are implementing changes that are intended to increase performance, you
    should provide the results of a simple performance benchmark exercise
    demonstrating the improvement. Especially if the changes make code less legible.

## :ballot_box_with_check: Checklist

- [x] :art: The changes implemented is consistent with the [julia style guide](https://docs.julialang.org/en/stable/manual/style-guide/).
- [x] :blue_book: I have updated and added relevant docstrings, in a manner consistent with the [documentation styleguide](https://docs.julialang.org/en/stable/manual/documentation/).
- [x] :blue_book: I have added or updated relevant user and developer manuals/documentation in `docs/src/`.
- [x] :ok: There are unit tests that cover the code changes I have made.
- [x] :ok: The unit tests cover my code changes AND they pass.
- [ ] :pencil: I have added an entry to the `[UNRELEASED]` section of the manually curated `CHANGELOG.md` file for this repository.
- [x] :ok: All changes should be compatible with the latest stable version of Julia.
- [x] :thought_balloon: I have commented liberally for any complex pieces of internal code.

jangevaare commented:
Thank you for the PR, going through this a bit now. Couple initial comments:

Would it be better to add a scale to each type, such that it is calculated once during construction rather than each time the `P()` or `Q()` functions are used?

I see `;scale=false` is added as a keyword argument to some functions but it doesn't seem to be used - everything is just scaled. I do think having both options (scaled vs. non-scaled) is important, what is the default behaviour can be debated though.

kgori commented:
Hi, thanks for the feedback. I'd like to respond to your comments:

Calculating the scale once during construction is a good idea. It would be valid as long as the model parameters don't change, which they won't as I think they (or the struct) are immutable. Along these lines I think Q could also be calculated once during construction. As far as this PR goes, I can move the scale calculation into the constructor, but I think doing the same with the Q might be work for another day.

I wasn't sure if I should make scaling optional or the default. In the last commit of this PR I made it the default to always scale the matrices, but in the previous commit there is a keyword `;scale=false` argument to `P()` and `Q()` that controls whether scaling is applied. It looks like your preference is for having the control, so I can roll back the last commit.

What do you think of this plan? 1) Roll back the commit so scaling is optional (default: scaling=true?), and 2) move the scaling calculation into the constructor. Is it better to do this within this pull request, or drop it and start another so there's less commit-pollution?

jangevaare commented:
What about just something just like in the link below. The generic scale method and scaled Q method could be overloaded for individual substitution models, though I'm not sure there's a lot to gain there computationally.

https://github.com/BioJulia/SubstitutionModels.jl/commit/4951b1b582041d7b980572696ecb0fad2b67a9a6#diff-240dae511ec080da77ea6ccd9d8acf48

kgori committed:
Add scaling feature to P and Q matrix calculations

kgori commented:
I like your suggestion to do this with overloaded functions, it's much simpler than my approach. I've dropped my earlier commits and added scaled versions of `Q()` and `P()` using your method.

jangevaare commented on src/nucleic_acid/nucleic_acid.jl:
```diff
@@ -49,6 +66,24 @@ end
 end
 
 
+function P_generic(mod::NASM, t::Float64, scaled::Bool)
```
If you leave this as P_generic(mod::NASM, t, scaled::Bool), Julia should figure out what to do. It would then make the implementation for arrays of time unnecessary in lines 78-84

jangevaare commented on src/nucleic_acid/nucleic_acid.jl:
```diff
@@ -41,6 +41,23 @@ Q_{T, A} & Q_{T, C} & Q_{T, G} & Q_{T, T} \\end{bmatrix}\$
                    β * πT, α * πT, γ * πT, -(β * πA + α * πC + γ * πG))
 end
 
+function Q(mod::NASM, scaled::Bool)
```
Pedantic, but can we use `scale` as the argument name. With `scaled` I find if it's ambigious whether it is an indication if something has already been scaled vs. an instruction for the function to perform the scaling. To me it is a bit more clear with an argument `scale` that it is instructive.

kgori replied:
I have no problem changing the name to `scale`, but it shadows the name of the function `scale`, leading to an error when `P_generic` calls `scale` on line 72:

    ERROR: MethodError: objects of type Bool are not callable

I can change the call on line 72 to `scale_generic(mod)` or `SubstitutionModels.scale(mod)` to prevent this.

jangevaare replied:
Right, of course. We could use `_scale()` as the function, and not export as another option. It seems to me that calculating the scale outside of the P and Q functions now would not be common.

jangevaare commented:
Works, for me, thanks!

jangevaare commented:
If you could just add a sentence to the docstrings for `P()`, and `Q()` functions about this, then squash these commits down to something like "Add scaling feature to P and Q matrix calculations", I will merge and push a new release.

kgori commented:
I changed the docs and squashed the commits, so I think this is ready to go.

jangevaare commented on src/nucleic_acid/k80/absolute.jl:
```diff
@@ -41,8 +41,8 @@ end
 
-  e₁ = 0.25 * exp(-4 * β * t)
-  e₂ = 0.5 * exp(-2 * (α + β) * t)
+  e₁ = 0.25 * exp(-β * t)
```
Is this change meant to be left in? Or was there an error with how K80 model P matrices were specified?

kgori replied:
Not an error, but they were already set up to do scaling. This change makes them act the same as the other models, only scale with `P(mod, true)`.

jangevaare commented on src/nucleic_acid/k80/relative.jl:
```diff
@@ -38,8 +38,8 @@ end
 
-  e₁ = 0.25 * exp(-4 * t/(κ + 2))
-  e₂ = 0.5 * exp(-2 * t * (κ + 1)/(κ + 2))
+  e₁ = 0.25 * exp(-t)
```
Same comment as left on k80/absolute.jl

jangevaare commented:
FYI - I'm going to tag this as v0.2.3, with a v0.3.0 coming soon after with changes to use project.toml
